### ISSUE TITLE
Question: Does alloydb-auth-proxy support the use of GKE workload identity

### ISSUE THREAD

**Issue by adamstrawson** [GoogleCloudPlatform/alloydb-auth-proxy#163]
## Question
Does `alloydb-auth-proxy` support the use of GKE workload identity when used with ServiceAccounts?

I'm in the process of setting up alloydb, and the plan is to run `alloydb-auth-proxy` as a sidecar container, similar to what we currently do with `cloud-sql-proxy`.

However, when using ServiceAccounts with workload identity, `alloydb-auth-proxy` fails to get the instance IP address. 

```
2022/11/02 17:55:53 [projects/{project}/locations/{region}/clusters/{cluster}/instances/{instance}] failed to connect to instance: failed to get instance IP address: Refresh error: failed to get instance metadata (instance URI = "{project}/{region}/{cluster}/{instance}"): Get "https://alloydb.googleapis.com/v1beta/projects/{project}/locations/{region}/clusters/{cluster}/instances/{instance}/connectionInfo": compute: Received 403 `Unable to generate access token; IAM returned 403 Forbidden: The caller does not have permission
```

No matter what role I include on the IAM Service Account, I can't get passed this error. 

Example Kubernetes resources below;

```yaml
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: alloydb-proxy
  namespace: platform
  annotations:
    iam.gke.io/gcp-service-account: alloydb-proxy@{project}.iam.gserviceaccount.com
```

```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: alloydb-proxy
  namespace: platform
spec:
  replicas: 1
    spec:
      serviceAccountName: alloydb-proxy
      containers:
      - name: alloydb-proxy
        image: gcr.io/alloydb-connectors/alloydb-auth-proxy:0.6.0
        ports:
        - containerPort: 5432
        command:
        - /alloydb-auth-proxy
        - {instance_uri}
```

The ServiceAccount `alloydb-proxy@{project}.iam.gserviceaccount.com` has the appropriate workload identity permissions (which we already use across the cluster and works fine, such as `cloud-sql-proxy`), originally the ServiceAccount had the role `role/alloydb.client`, but I've attempted a few variations around token creator etc, with no luck

**Comment by enocom** [comment]
Workload identity is supported by the proxy as it uses application default credentials under the hood.

Have you verified that your workload is in fact running with the service account you expect to use?

Cf. https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#verify_the_setup.

**Comment by adamstrawson** [comment]
HI @enocom, thanks for coming back to me so quickly!

I've verified that our workload is running with the service account correctly

```
curl -H "Metadata-Flavor: Google" http://169.254.169.254/computeMetadata/v1/instance/service-accounts/default/email
alloydb-proxy@{project}.iam.gserviceaccount.com
```

**Comment by enocom** [comment]
Let ms try this out myself. I would expect workload identity to be supported, but will confirm just in case and report back here.

**Comment by enocom** [comment]
So here's what I did:

1. Used a fresh GKE cluster with Workload Identity Enabled
2. Configured applications to use workload identity based [on the docs](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#authenticating_to)
3. Added roles/alloydb.client to my workload identity
4. Deployed a toy app that reads the time from the AlloyDB cluster
5. Saw the time come back from the cluster

So in short, the proxy does support workload identity. At a glance, it looks like your setup is correct, but there must be something incorrect in there.

You might try taking Kubernetes out of the picture and running the Proxy locally with a service account key, to see if you get the same 403 error and go from there.

**Comment by adamstrawson** [comment]
Hi @enocom Thanks for the support on this, found the issue, completely down to user error on my part! 

Typo on the namespace for the principle.

Really appreciate the support though!

**Comment by enocom** [comment]
So happy to hear it. Workload Identity is easy to trip over with all the configuration.
